### PR TITLE
1291960 - Fix restarting goferd while being updated by RPM.

### PR DESCRIPTION
**Pull request by jortel**
https://bugzilla.redhat.com/show_bug.cgi?id=1291960
https://bugzilla.redhat.com/show_bug.cgi?id=1361268

This a possible solution.

Assuming the `%post` in the katello-agent RPM has been updated to write `/tmp/katello-agent-restart`, the plugin will detect the file and restart `goferd` when the plugin is idle.  Admittedly, this is a little hacky but given the near term replacement of this technology, the simpler the better.

The `@action(seconds=5)` should probably be longer (perhaps minutes=5) but left it this way for testing.

A potential alternative option could be to change the `%post` to use the `at` command to delay execute a shell script to restart the service when the `/var/lib/gofer/messaging/pending/katelloagent` directory is empty (things are idle) just as the python code here does.  But, this requires the `atd` service to be running.

### CODE REVIEW

**Comment by jortel** [comment]
Unit tests will be added pending approval of the approach.

File: src/katello/agent/katelloplugin.py [hunk 1]
```diff
@@ -209,6 +210,70 @@ def validate_registration():
...
+          2. Restart
+          (restart failed)
+          3. Touch (create) the RESTART_FILE
+        """
+        os.unlink(AgentRestart.RESTART_FILE)
+        try:
+            p = Popen(AgentRestart.COMMAND, shell=True)
+            p.wait()
+        finally:
+            AgentRestart.touch()
```
Comment (jlsherrill) [inline src/katello/agent/katelloplugin.py]: why does it touch the RESTART_FILE after a restart?

Comment (jortel) [inline src/katello/agent/katelloplugin.py]: If the `Popen()` succeeds and goferd is restarted, the `finally:` will never get executed.  But, if the restart fails or raises an exception, the file get's re-created so the request for restart is preserved.  Question though:  What should we do if the restart fails?  This logic ^ is intended to retry indefinitely which may not make sense. 

Options:
- retry indefinitely
- retry a fixed number of times.
- only try once.

Thoughts?

Comment (zjhuntin) [inline src/katello/agent/katelloplugin.py]: I think that the fire-and-forget way would be the best. I think anything that might fix the issue would happen after one retry, and continuing indefinitely sounds like it'd be less desirable, as wouldn't stop the cycle at any point.

Comment (jlsherrill) [inline src/katello/agent/katelloplugin.py]: Yeah, i'd tend to agree.  Can you think of any situation where the process wouldn't die properly?

Comment (jortel) [inline src/katello/agent/katelloplugin.py]: nope

**Comment by jortel** [comment]
PR updated to *fire-and-forget* as discussed.

**Comment by zjhuntin** [comment]
I have tested this with katello and some scratch builds of katello-agent, and it works great. I was able to update katello-agent through the web-ui. Upon a successful update I also confirmed that goferd was restarted by making sure the process had a different PID than before the update.

Thanks @jortel ! ACK

**Comment by zjhuntin** [comment]
@jortel I've made a redmine issue for this issue: http://projects.theforeman.org/issues/18187 Would you mind changing the commit message to Fixes #18187 - <Message>. Also we approve of the approach and would love those tests.

**Comment by jortel** [comment]
Tests added with 100% coverage and commit changed as requested.

I also changed the interval to: minutes=3

**Comment by jlsherrill** [comment]
ACK thanks @jortel

**Pull request merged**